src/bom/chainsop.rs

```diff
@@ -0,0 +1,872 @@
+    /// still be created in the normal temporary directory location.
+    #[inline]
+    pub fn set_dir<T>(&mut self, in_dir: T) -> &SubProcOperation
+    where T: AsRef<Path>
+    {
+        self.in_dir = Some(in_dir.as_ref().to_path_buf());
+        self
+    }
+
+    /// Executes this command in a subprocess in the specified directory.  The
+    /// input and output files will be determined and added to the command-line
+    /// as indicated by their FileSpec values.  The successful result specifies
+    /// the output file written (if any).
+    pub fn execute(&self, cwd: &Path) -> anyhow::Result<SubProcFile>
+    {
+        let mut args = self.args.clone();
+        let outfile;
+        if self.emit_output_file_first() {
+            outfile = self.out_file.setup_file(
+                &mut args,
+                || Err(anyhow::Error::new(
+                    SubProcError::ErrorMissingFile(String::from(&self.cmd),
+                                                   String::from("output")))))?;
+            self.inp_file.setup_file(
+                &mut args,
+                || Err(anyhow::Error::new(
+                    SubProcError::ErrorMissingFile(String::from(&self.cmd),
+                                                   String::from("input")))))?;
+        } else {
+            self.inp_file.setup_file(
+                &mut args,
+                || Err(anyhow::Error::new(
+                    SubProcError::ErrorMissingFile(String::from(&self.cmd),
+                                                   String::from("input")))))?;
+            outfile = self.out_file.setup_file(
+                &mut args,
+                || Err(anyhow::Error::new(
+                    SubProcError::ErrorMissingFile(String::from(&self.cmd),
+                                                   String::from("output")))))?;
+        }
+        self.run_cmd(cwd, outfile, args)
+    }
+    // Output option arguments before positional arguments because some command's
+    // parsers are limited in this way.  This function returns true if the output
+    // file should be specified before the input file; the normal order is input
+    // file and then output file (e.g. "cp inpfile outfile").
+    fn emit_output_file_first(&self) -> bool
+    {
+        if let FileSpec::Option(_, _) = self.out_file {
+            if let FileSpec::Append(_) = self.inp_file {
+                true
+            } else {
+                false
+            }
+        } else {
+            false
+        }
+    }
+
+    // After the files are setup, this performs the actual run
+    fn run_cmd(&self, cwd: &Path, outfile : SubProcFile, args : Vec<OsString>)
+               -> anyhow::Result<SubProcFile>
+    {
+        match &self.cmd {
+            Operation::Execute(cmd) => {
+                match process::Command::new(&cmd)
+                .args(&args)
+                .current_dir(self.in_dir.as_ref().unwrap_or(&cwd.to_path_buf()))
```

Comment:
I think the `&cwd` is redundant and could just be `cwd`, as it is already a ref.  Also consider using `unwrap_or_else` here

Reply:
Changed, thanks for catching these optimizations.